Batteries/Control/LawfulMonadState.lean

```diff
@@ -0,0 +1,235 @@
+    (do set (m := m) s; f <$> set s') = (do f <$> set s') := by
+  simp [map_eq_pure_bind, ← bind_assoc, -bind_pure_comp]
+
+section modify
+
+theorem modifyGetThe_eq (f : σ → α × σ) :
+    modifyGetThe σ (m := m) f = do let z ← f <$> get; set z.2; return z.1 := modifyGet_eq f
+
+theorem modify_eq (f : σ → σ) :
+    modify (m := m) f = (do set (f (← get))) := by simp [modify, modifyGet_eq]
+
+theorem modifyThe_eq (f : σ → σ) :
+    modifyThe σ (m := m) f = (do set (f (← get))) := modify_eq f
+
+theorem getModify_eq (f : σ → σ) :
+    getModify (m := m) f = do let s ← get; set (f s); return s := by
+  rw [getModify, modifyGet_eq, bind_map_left]
+
+/-- Version of `modifyGet_eq` that preserves an call to `modify`. -/
+theorem modifyGet_eq' (f : σ → α × σ) :
+    modifyGet (m := m) f = do let s ← get; modify (Prod.snd ∘ f); return (f s).fst := by
+  simp [modify_eq, modifyGet_eq]
+
+@[simp] theorem modify_id : modify (m := m) id = pure PUnit.unit := by
+  simp [modify_eq]
+
+@[simp] theorem getModify_id : getModify (m := m) id = get := by
+  simp [getModify_eq]
+
+@[simp] theorem set_bind_modify (s : σ) (f : σ → σ) :
+    (do set (m := m) s; modify f) = set (f s) := by simp [modify_eq]
+
+@[simp] theorem set_bind_modify_bind (s : σ) (f : σ → σ) (mx : PUnit → m α) :
+    (do set s; let u ← modify f; mx u) = (do set (f s); mx PUnit.unit) := by
+  simp [modify_eq, ← bind_assoc]
+
+@[simp] theorem set_bind_modifyGet (s : σ) (f : σ → α × σ) :
+    (do set (m := m) s; modifyGet f) = (do set (f s).2; return (f s).1) := by simp [modifyGet_eq]
+
+@[simp] theorem set_bind_modifyGet_bind (s : σ) (f : σ → α × σ) (mx : α → m β) :
+    (do set s; let x ← modifyGet f; mx x) = (do set (f s).2; mx (f s).1) := by simp [modifyGet_eq]
+
+@[simp] theorem set_bind_getModify (s : σ) (f : σ → σ) :
+    (do set (m := m) s; getModify f) = (do set (f s); return s) := by simp [getModify_eq]
+
+@[simp] theorem set_bind_getModify_bind (s : σ) (f : σ → σ) (mx : σ → m α) :
+    (do set s; let x ← getModify f; mx x) = (do set (f s); mx s) := by
+  simp [getModify_eq, ← bind_assoc]
+
+@[simp] theorem modify_bind_modify (f g : σ → σ) :
+    (do modify (m := m) f; modify g) = modify (g ∘ f) := by simp [modify_eq]
+
+@[simp] theorem modify_bind_modifyGet (f : σ → σ) (g : σ → α × σ) :
+    (do modify (m := m) f; modifyGet g) = modifyGet (g ∘ f) := by
+  simp [modify_eq, modifyGet_eq]
+
+@[simp] theorem getModify_bind_modify (f : σ → σ) (g : σ → σ → σ) :
+    (do let s ← getModify (m := m) f; modify (g s)) =
+      (do let s ← get; modify (g s ∘ f)) := by
+  simp [modify_eq, getModify_eq]
+
+theorem modify_comm_of_comp_comm {f g : σ → σ} (h : f ∘ g = g ∘ f) :
+    (do modify (m := m) f; modify g) = (do modify (m := m) g; modify f) := by
+  simp [modify_bind_modify, h]
+
+theorem modify_bind_get (f : σ → σ) :
+    (do modify (m := m) f; get) = (do let s ← get; modify f; return (f s)) := by
+  simp [modify_eq]
+
+end modify
+
+end LawfulMonadStateOf
+
+namespace StateT
```

Comment:
Namespaces have semantic meaning that is not used here. A comment would have the same effect,